### PR TITLE
fix: escape unsafe patterns in elements, comment nodes and processing instructions

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit updates the logic to espace unsafe patterns:

- for elements that can have raw text: all matching closing tags are escaped
- for comment nodes: all closing comment tags are escaped
- for processing instructions: all `>` symbols are escaped

### CODE REVIEW

File: lib/NodeUtils.js [hunk 1]
```diff
@@ -98,6 +98,49 @@ function attrname(a) {
...
+    } else {
+      if (rawText.substr(i, 4) === '<!--') {
+        // We came across an open comment tag,
+        // set the flag and proceed to the next char.
+        isInCommentBlock = true;
+      } else {
+        // We are not in comment block, check if we have
+        // a lookup pattern in front of us. Compare against
+        // a lowercased content to avoid case sensitivity.
+        if (lowercasedRawText.substr(i, pattern.length) === pattern) {
```
Comment (alan-agius4) [inline lib/NodeUtils.js:133]: `substr` is deprecated, use `slice` or `substring` instead

File: lib/NodeUtils.js [hunk 2]
```diff
@@ -98,6 +98,49 @@ function attrname(a) {
...
+/**
+ * Escapes matching closing tag in a raw text.
+ *
+ * For example, given `<style>#text(</style><script></script>)</style>`,
+ * the parent tag would by "style" and the raw text is
+ * "</style><script></script>". If we come across a matching closing tag
+ * (in out case `</style>`) - replace `<` with `&lt;` to avoid unexpected
+ * and unsafe behavior after de-serialization.
+ */
+function escapeMatchingClosingTag(rawText, parentTag) {
```
Comment (alan-agius4) [inline lib/NodeUtils.js:110]: Nit: from a performance point of view would it be more optimal to do an initial scan to check for either a closing tag `</` or the closing pattern `'</' + parentTag.toLowerCase() + '>';`?

File: lib/NodeUtils.js [hunk 3]
```diff
@@ -116,6 +116,82 @@ function attrname(a) {
...
+const CLOSING_COMMENT_REGEXP_GLOBAL = /--!?>/g;
+
+/**
+ * Escapes closing comment tag in a comment content.
+ *
+ * For example, given `#comment('-->')`, the content of a comment would be
+ * updated to `--&gt;` to avoid unexpected and unsafe behavior after
+ * de-serialization.
+ */
+function escapeClosingCommentTag(rawContent) {
```
Comment (securityMB) [inline lib/NodeUtils.js:163]: There's an off-by-one error in this function. 

```
escapeClosingCommentTag("asd-->a<script>")
// returns 'asd-->&gt;<script>'
```

The comment end still persists and erroneously the next character (letter `"a"`) is being replaced with `&gt;` instead.

I also wonder whether this function can be replaced with just a `rawContent.replaceAll` but I might be missing some edge cases here.

Comment (AndrewKushnir) [reply lib/NodeUtils.js:162]: @securityMB great catch, thanks for looking into this! 👍

I've applied the changes in https://github.com/angular/domino/pull/18/commits/9bf2b98233b61d1c92eca4fb4b78afb3e9786572 and also extended tests to verify HTML output (some tests previously only ran a check in a browser).

I think using `replaceAll` is a good idea, please let me know if you see any edge-cases.

**Review by securityMB — APPROVED** [review]
LGTM. Thanks for putting this up!

**Comment by AndrewKushnir** [comment]
@securityMB thanks for the review 👍

FYI I've used your recommendation and also added tests for escaping util functions (in https://github.com/angular/domino/pull/18/commits/8f25adc7540cca5c9ee6de84d43ce88f44c5a3b4), so it's easier to extend the coverage if needed.

File: lib/NodeUtils.js [hunk 4]
```diff
@@ -116,6 +123,65 @@ function attrname(a) {
...
+    // Skip over the `<` char, since we've replaced it above.
+    // Start the next iteration with a char following the `<` one.
+    lastIndex = match.index + 1;
+  }
+  // Add remaining contents of a string after the last `<` char.
+  if (lastIndex < rawText.length) {
+    result += rawText.substring(lastIndex);
+  }
+  return result;
+}
```
Comment (alan-agius4) [inline lib/NodeUtils.js:158]: On seconds thoughts this entire method can be simplified and written in a more performant way
```suggestion
function escapeMatchingClosingTag(rawText, parentTag) {
  if (!rawText.toLowerCase().includes('</'+ parentTag)) {
    return rawText; // fast path
  }

  const result = [...rawText];
  const matches = rawText.matchAll('</' + parentTag, 'ig');
  for (const match of matches) {
    result[match.index] = '&lt;';
  }

  return result.join('');
}

```

**Review by alan-agius4 — APPROVED** [review]
LGTM, just some performance related observations. (Totally optional)

**Comment by AndrewKushnir** [comment]
@alan-agius4 thanks for reviewing, I've applied the necessary changes.
